modules/xfem/src/userobjects/InterfaceMeshCut3DUserObject.C

```diff
@@ -0,0 +1,406 @@
+                                                   std::vector<Xfem::CutFace> & cut_faces) const
+{
+  bool elem_cut = false;
+
+  if (elem->dim() != 3)
+    mooseError("The structural mesh to be cut by a surface mesh must be 3D!");
+
+  for (unsigned int i = 0; i < elem->n_sides(); ++i)
+  {
+    // This returns the lowest-order type of side.
+    std::unique_ptr<const Elem> curr_side = elem->side_ptr(i);
+    if (curr_side->dim() != 2)
+      mooseError("In cutElementByGeometry dimension of side must be 2, but it is ",
+                 curr_side->dim());
+    unsigned int n_edges = curr_side->n_sides();
+
+    std::vector<unsigned int> cut_edges;
+    std::vector<Real> cut_pos;
+
+    for (unsigned int j = 0; j < n_edges; j++)
+    {
+      // This returns the lowest-order type of side.
+      std::unique_ptr<const Elem> curr_edge = curr_side->side_ptr(j);
+      if (curr_edge->type() != EDGE2)
+        mooseError("In cutElementByGeometry face edge must be EDGE2, but type is: ",
+                   libMesh::Utility::enum_to_string(curr_edge->type()),
+                   " base element type is: ",
+                   libMesh::Utility::enum_to_string(elem->type()));
+      const Node * node1 = curr_edge->node_ptr(0);
+      const Node * node2 = curr_edge->node_ptr(1);
+
+      for (const auto & cut_elem : _cutter_mesh->element_ptr_range())
+      {
+        std::vector<Point> vertices;
+
+        for (auto & node : cut_elem->node_ref_range())
+        {
+          Point & this_point = node;
+          vertices.push_back(this_point);
+        }
+
+        Point intersection;
+        if (intersectWithEdge(*node1, *node2, vertices, intersection) &&
+            std::find(cut_edges.begin(), cut_edges.end(), j) == cut_edges.end())
+        {
+          cut_edges.push_back(j);
+          cut_pos.emplace_back(getRelativePosition(*node1, *node2, intersection));
+        }
+      }
+    }
+
+    // if two edges of an element are cut, it is considered as an element being cut
+    if (cut_edges.size() == 2)
+    {
+      elem_cut = true;
+      Xfem::CutFace mycut;
+      mycut._face_id = i;
+      mycut._face_edge.push_back(cut_edges[0]);
+      mycut._face_edge.push_back(cut_edges[1]);
+      mycut._position.push_back(cut_pos[0]);
+      mycut._position.push_back(cut_pos[1]);
+      cut_faces.push_back(mycut);
+    }
+  }
+
+  return elem_cut;
+}
+
+bool
+InterfaceMeshCut3DUserObject::cutFragmentByGeometry(
+    std::vector<std::vector<Point>> & /*frag_edges*/,
+    std::vector<Xfem::CutEdge> & /*cut_edges*/) const
+{
+  mooseError("invalid method for 3D mesh cutting");
+  return false;
+}
+
+bool
+InterfaceMeshCut3DUserObject::cutFragmentByGeometry(
+    std::vector<std::vector<Point>> & /*frag_faces*/,
+    std::vector<Xfem::CutFace> & /*cut_faces*/) const
+{
+  mooseError("cutFragmentByGeometry not yet implemented for 3D mesh cutting");
+  return false;
+}
+
+bool
+InterfaceMeshCut3DUserObject::intersectWithEdge(const Point & p1,
```

Comment:
See comments above. I think this method is also something that could be useful in general. So putting it in a utility class would be better.

Reply:
done